### PR TITLE
Avoid fetching image dimensions when for internal image.

### LINKED ISSUE #5115: Avoid fetching image dimensions over HTTP when possible
## Feature description

Fetching an image's dimensions via `FasterImage` over HTTP is resource-intensive. When possible, the AMP plugin should be fetching the dimensions by examining the file from the filesystem. 

We should check to see whether it is more expensive to obtain the dimensions of the image at runtime from the filesystem versus obtaining the dimensions of the image via a transient.

Additionally, when the image actually contains the dimensions in the filename (e.g. `foo-100x200.png`) then this should serve indicate the dimensions without necessitating the request.

Related: #4838.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by dhaval-parekh**
## Summary

Fixes #5115 

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]: I'm not sure we should try to obtain the attachment post. For example, if an image URL looks like `https://example.com/wp-content/uploads/2021/04/American_bison_k5680-1-1024x668.jpg`, it will never match an attachment image because this is a generated image size. 

What's more, if the image URL does end in `-\d+x\d+\.\w+` should this bypass the need to further look up the dimensions? Could the dimensions then just be obtained from the filename itself?

Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php]: In other words, the logic you have at the  end of this function should perhaps come first, even for images that are linked to from external domains.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]: Why `strtolower`?

Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php]: We'd want to make sure the string appears at the beginning:

```suggestion
			if ( 0 !== strpos( strtolower( trim( $url ) ), $base_upload_uri ) ) {
```

Comment (dhaval-parekh) [reply includes/utils/class-amp-image-dimension-extractor.php]: Addressed.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (dhaval-parekh) [reply includes/utils/class-amp-image-dimension-extractor.php]: I have updated the PR.

I kept the logic of getting image dimensions from the URL at last because of concerns for those images that have dimensions in the name but not actual dimensions. But this is not applicable now. Since WordPress add a suffix or those images.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]: I think this can be simplified:

```suggestion
			// Get image dimension from file name.
			$path = wp_parse_url( $url, PHP_URL_PATH );
			if ( empty( $path ) ) {
				continue;
			}
			
			if ( preg_match( '/-(?<width>\d+)x(?<height>\d+)\.\w+$/', $path, $possible_dimension ) ) {
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]: The logic for obtaining the cached dimensions transient is now duplicated in this class. There should be a helper method `get_cached_dimensions_transient_name` that contains the duplicated logic.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]: The `$url` could have query parameters here. So those should get stripped off.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]:
```suggestion
			if ( 0 !== strpos( $url, $base_upload_uri ) ) {
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php]: Also, instead of doing `str_replace()` I suggest using `substr( $url, strlen( $base_upload_uri ) )` since we know we want to strip off a certain number of characters from the beginning as discovered above.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php:216]: This should make sure that the `$url` is not malicious in any way, like using `bad/../../path.evil`.

```suggestion
	if ( 0 !== validate_url( $attached_path ) ) {
		continue;
	}

```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]:
```suggestion
				} elseif ( function_exists( 'getimagesize' ) ) {
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php]:
```suggestion
					$imagesize = @getimagesize( $image_file ); // phpcs:ignore WordPress.PHP.NoSilencedErrors
```

File: tests/php/test-amp-image-dimension-extractor.php
Comment (westonruter) [inline tests/php/test-amp-image-dimension-extractor.php]:
```suggestion
	 * @covers \AMP_Image_Dimension_Extractor::extract_by_filename_or_filesystem()
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [inline includes/utils/class-amp-image-dimension-extractor.php:160]: This should have a `dataSet` to test with `wp_using_ext_object_cache()` being both `true` and `false`.

File: tests/php/test-amp-image-dimension-extractor.php
Comment (westonruter) [inline tests/php/test-amp-image-dimension-extractor.php]:
```suggestion
		// Check after removing attachment metadata, to check fallback to filesystem.
		wp_update_attachment_metadata( $attachment_id, [] );
```

File: tests/php/test-amp-image-dimension-extractor.php
Comment (westonruter) [inline tests/php/test-amp-image-dimension-extractor.php]: To facilitate maintenance, I suggest that these two be combined into one array, like:

```php
$data = [
    $full_image[0] => [
        'input' => [],
        'expected' => [
            'width'  => $full_image[1],
            'height' => $full_image[2],
        ],
    ],
    ...
]
```

Then you can extract the `$expected` and `$input` via `array_map()`.

**Review by westonruter — CHANGES_REQUESTED** [review]

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php:160]: Sorry, I meant to leave this comment on the test.

**Review by westonruter — APPROVED** [review]
The changes check out to me. Good work.

I want to get a second review from @pierlon as well since this touches on some fundamental aspects of how the plugin has worked for a long time.

**Comment by westonruter** [comment]
Note this will specifically reduce data being stored in options as is a valid complaint about the plugin: https://wordpress.org/support/topic/this-plugin-inserted-a-lot-of-data-in-the-wp_options-table-2/

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (pierlon) [inline includes/utils/class-amp-image-dimension-extractor.php]: Although unlikely, there may be a site or two that are using this filter, and if so it would be likely they are running at the default priority 10. I'd recommend setting the priority of this callback to something > 10 especially since the filtered value would affect the behavior of the callback.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (pierlon) [inline includes/utils/class-amp-image-dimension-extractor.php:215]: I think these should be the first set of checks before parsing the URL for dimensions.

**Review by pierlon — CHANGES_REQUESTED** [review]

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php:215]: I'm not sure because the function expects a file path as input, not a URL. So I think we should make sure it looks like a file path, or else this could break if `validate_file()` changes its behavior in the future which isn't compatible with URLs.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (pierlon) [inline includes/utils/class-amp-image-dimension-extractor.php]:
```suggestion
			if ( ! in_array( strtolower( $matches['ext'] ), $image_ext_types, true ) ) {
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (pierlon) [reply includes/utils/class-amp-image-dimension-extractor.php:215]: I meant moving both checks. So it would look like:

```diff
diff --git a/includes/utils/class-amp-image-dimension-extractor.php b/includes/utils/class-amp-image-dimension-extractor.php
--- a/includes/utils/class-amp-image-dimension-extractor.php	(revision 35310f1ff829be5a145d14e72af27a28b2cfc0c6)
+++ b/includes/utils/class-amp-image-dimension-extractor.php	(date 1628092492414)
@@ -184,6 +184,17 @@
 
 			$url_without_query_fragment = strtok( $url, '?#' );
 
+			// Verify that the URL is for an uploaded file.
+			if ( 0 !== strpos( $url_without_query_fragment, $base_upload_uri ) ) {
+				continue;
+			}
+			$upload_relative_path = substr( $url_without_query_fragment, strlen( $base_upload_uri ) );
+
+			// Bail if the URL contains relative paths.
+			if ( 0 !== validate_file( $upload_relative_path ) ) {
+				continue;
+			}
+
 			// Parse info out of the URL, including the file extension and (optionally) the dimensions.
 			if ( ! preg_match( '/(?:-(?<width>[1-9][0-9]*)x(?<height>[1-9][0-9]*))?\.(?<ext>\w+)$/', $url_without_query_fragment, $matches ) ) {
 				continue;
@@ -202,17 +213,6 @@
 				];
 				continue;
 			}
-
-			// Verify that the URL is for an uploaded file.
-			if ( 0 !== strpos( $url_without_query_fragment, $base_upload_uri ) ) {
-				continue;
-			}
-			$upload_relative_path = substr( $url_without_query_fragment, strlen( $base_upload_uri ) );
-
-			// Bail if the URL contains relative paths.
-			if ( 0 !== validate_file( $upload_relative_path ) ) {
-				continue;
-			}
 
 			// Get image dimension from file system.
 			$image_file = $base_upload_dir . $upload_relative_path;
```

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php]: How about 100?

```suggestion
		add_filter( 'amp_extract_image_dimensions_batch', [ __CLASS__, 'extract_by_filename_or_filesystem' ], 100 );
```

Comment (pierlon) [reply includes/utils/class-amp-image-dimension-extractor.php]: Sounds good 👍.

File: includes/utils/class-amp-image-dimension-extractor.php
Comment (westonruter) [reply includes/utils/class-amp-image-dimension-extractor.php:215]: Ah, I just realized that this needs to be ordered in this way. The reason is that we will use the dimensions in the filename for remote files which have URLs that don't start with `$base_upload_uri`. 

So the order is correct as it stands.

Comment (pierlon) [reply includes/utils/class-amp-image-dimension-extractor.php:215]: Ah I see, makes sense 👍

**Review by pierlon — APPROVED** [review]

**Pull request merged**
